replay/models/nmslib_hnsw.py

```diff
@@ -0,0 +1,646 @@
+                                (item_idxs[non_seen_item_indexes])[:k],
+                                (distances[non_seen_item_indexes])[:k],
+                            )
+                        )
+
+                    pd_res = pd.DataFrame(
+                        neighbours_filtered, columns=["item_idx", "distance"]
+                    )
+
+                    return pd_res
+
+            else:
+
+                @pandas_udf(return_type)
+                def infer_index(vectors: pd.Series) -> pd.DataFrame:
+                    index_file_manager = index_file_manager_broadcast.value
+                    index = index_file_manager.index
+
+                    neighbours = index.knnQueryBatch(
+                        np.stack(vectors.values),
+                        k=k,
+                        num_threads=1,
+                    )
+                    pd_res = pd.DataFrame(
+                        neighbours, columns=["item_idx", "distance"]
+                    )
+
+                    return pd_res
+
+        cols = []
+        if index_type == "sparse":
+            cols += ["user_idx", "vector_items", "vector_relevances"]
+        else:
+            cols.append(features_col)
+        if filter_seen_items:
+            cols = cols + ["num_items", "seen_item_idxs"]
+
+        res = user_vectors.select(
+            "user_idx",
+            infer_index(*cols).alias("neighbours"),
+        )
+
+        res = self._unpack_infer_struct(res)
+
+        return res
+
+    def _save_nmslib_hnsw_index(self, path, sparse=False):
+        """Method save (copy) index from hdfs (or local) to `path` directory.
+        `path` can be an hdfs path or a local path.
+
+        Args:
+            path (_type_): directory where to dump (copy) the index
+        """
+
+        params = self._nmslib_hnsw_params
+
+        if params["build_index_on"] == "executor":
+            index_path = params["index_path"]
+        elif params["build_index_on"] == "driver":
+            index_path = SparkFiles.get(
+                f"{INDEX_FILENAME}_{self._spark_index_file_uid}"
+            )
+        else:
+            raise ValueError("Unknown 'build_index_on' param.")
+
+        source = get_filesystem(index_path)
+        target = get_filesystem(path)
+        self.logger.debug("Index file coping from '%s' to '%s'", index_path, path)
+
+        source_paths = []
+        target_paths = []
+        if sparse:
+            source_paths.append(source.path)
+            source_paths.append(source.path + ".dat")
+            index_file_target_path = os.path.join(target.path, INDEX_FILENAME)
+            target_paths.append(index_file_target_path)
+            target_paths.append(index_file_target_path + ".dat")
+        else:
+            source_paths.append(source.path)
+            index_file_target_path = os.path.join(target.path, INDEX_FILENAME)
+            target_paths.append(index_file_target_path)
+
+        if source.filesystem == FileSystem.HDFS:
```

Comment:
please revise this logic according to single responsibility principle, e.g through inheritance